### DISCUSSION TITLE
New Architecture Support Discussion

### DISCUSSION THREAD

**Discussion by hyochan** [hyochan/react-native-iap#2754] [General]
```
Announcing the Migration of react-native-iap to an Expo Module for Improved Maintenance and Compatibility
```

Hello React Native developers!

Today, I would like to share some thoughts after deep consideration regarding the maintenance of the `react-native-iap` module.

**TL;DR: We plan to migrate the `react-native-iap` module to an Expo module to align with the new industry standard, reduce maintenance costs, simplify the project, and improve compatibility.**

To briefly summarize, we plan to migrate the existing `react-native-iap` module to an Expo module instead of a pure React Native module. The reasons are as follows:

1. **New Industry Standard**: With the release of React Native version 0.74, it seems that the React Native official website encourages starting projects with Expo. The method to start a project with react-native-cli is somewhat hidden and requires some effort to find. This suggests that Expo projects are expected to become the industry standard in the future.

2. **Cost-Effective Maintenance**: Developing as an Expo module makes it much easier to manage native modules. Maintainers of React Native modules have had to invest a lot of time. For `react-native-iap` alone, we had to periodically update as React Native upgraded, and we also had to provide new features such as hooks. Additionally, we had to migrate according to native framework updates. For Android, there were many changes as billingclient upgraded. For iOS, we provided StoreKit and even supported StoreKit2. To provide these updates more efficiently, we also migrated to preferred languages for native development. For iOS, we moved from Objective-C to Swift, and for Android, from Java to Kotlin. The version updates of Kotlin alone often broke project dependencies (#2640, #2536, etc.). We also supported Expo config plugins to make it compatible with Expo (#2218). Now, we are at a point where we need to support the new architecture of React Native. After much effort to support the New Architecture, I believe that if we newly develop as an Expo module, we can greatly reduce these costs. Especially, Expo modules are "renderer-agnostic", meaning maintainers do not need to consider whether it is the New Architecture or Old Architecture when developing. Additionally, Expo modules can be easily installed with the `npx install-expo-modules@latest` command (https://docs.expo.dev/modules/overview).

3. **Simpler Project**: Following from reason 2, the current `react-native-iap` project has become very heavy and complex due to the effort to support multiple modules. We aim to simplify it based on the [design](https://docs.expo.dev/modules/overview/#design-considerations) of Expo Modules. Furthermore, we plan to provide easier and more convenient tutorials for all React Native developers starting with IAP.

4. **Better Compatibility**: By migrating to Expo Modules, we expect to resolve many compatibility issues we have faced over time.

These changes are expected to bring great advantages not only in developing the module but also in maintaining it, allowing many maintainers, including myself, to manage the module with much less burden.

As a potential downside, compatibility will follow Expo's support. Platforms not supported by Expo, such as Windows or Amazon, may not be supported. However, we can support them as Expo provides support.

I am sharing these thoughts to inform you of my current plan and to gather additional opinions from many of you. I expect to complete the work and release the expo-module in about 2 months.

Thank you for reading.

**Comment by MarkCSmith** [comment]
What would a migration of react-native-iap to an Expo module mean for existing React Native CLI projects?  Does using a mix of Expo and React Native modules work well?

**Reply by hyochan** [reply]
Expo module is compatible with react-native cli project.
You can simply install expo-modules with `npx install-expo-modules@latest` command.

Here is the [documentation](https://docs.expo.dev/bare/installing-expo-modules).

**Reply by MarkCSmith** [reply]
Thanks for your response and for pointing to the documentation.  I do not have any experience with Expo or Expo modules, so I will trust the maintainers of this project that this is a good choice.  The most important thing is to have an IAP module that works reliably and minimizes platform differences (Android + iOS).

**Comment by appideasDOTcom** [comment]
I agree with the move. I just wish that I needed it "in about 2 months" instead of today. :-)

I had significant issues attempting to update my project from Expo 50 to 51, so I'm stuck there until I can take some time to unravel and/or rework a couple things. I presume that updating to the Expo module version of this, when it's available, will require the latest version of Expo?

**Reply by hyochan** [reply]
This sounds crucial. I hope I can help! 

Regarding the question, "Will it require the latest version of Expo?" I'm not entirely sure, but I think Expo is quite well backward compatible if you are not using the bare workflow. They might deprecate some APIs, but I don't think these require crucial or urgent fixes.

**Comment by Jahanzeb009** [comment]
Excellent initiative!

**Comment by dev2xl** [comment]
Hello @hyochan do you have an estimated release date for this? I'm hoping to use this library after the migration. Thanks!

**Reply by hyochan** [reply]
I just posted in https://github.com/dooboolab-community/react-native-iap/discussions/2754#discussioncomment-10161345. I think I might finish the minimal version by mid to late August.

**Comment by mym0404** [comment]
It is unfortunate that there are currently three ways to write native libraries: old architecture, new architecture, and Expo module API.

Honestly, while I understand that Expo is recommended as a good way to start with React Native, I'm still not sure if libraries should also be written using the Expo module API. This may be because I have not yet seen examples of major libraries written with the Expo module API.

It seems the React Native team does not provide clear guidance on which method to stick to, and the scattered documentation only serves to confuse developers.

Of course, I believe that the API and developer experience (DX) perspective of the Expo module API are quite excellent.

Separate from that, although there is a one-line command to make Expo modules usable in bare React Native projects, I think this process could be as much of an obstacle for some as migrating from the old architecture to the new architecture.

Naturally, I understand that the decisions made by the creators of these libraries are most important to the libraries themselves.

https://github.com/react-native-community/discussions-and-proposals/blob/main/proposals/0759-react-native-frameworks.md#what-do-we-recommend-to-react-native-library-developers

**Reply by hyochan** [reply]
I'm having quite a good time using the Expo module API. The documentation is a bit weak, but as time goes by, I'm finding it really enjoyable. The modules are very well synced with the native side during development. Of course, I've faced some issues, such as functions that I previously implemented not working on another day. I found that I couldn't change some default configs that the Expo project provides to the module (such as iOS version in podspec) in order to make the module work concurrently when implementing a standalone library.

Also, I have a question about your comment below:

> "Separate from that, although there is a one-line command to make Expo modules usable in bare React Native projects, I think this process could be as much of an obstacle for some as migrating from the old architecture to the new architecture."

How could a one-line command cause issues with the old and new architecture?

**Reply by mym0404** [reply]
Migrating to a new architecture was just a metaphor.

I meant that integrating an Expo module package into a bare React Native project can sometimes be problematic.

This is because there shoudn't be a magic in the Expo module integration CLI. It needs to modify several configuration files in the native project, like AppDelegate.

If a beginner React Native developer blindly follows the setup guide and changes their native files, they might end up confused.

However, in my opinion, this shouldn’t be a problem. Never mind!

**Comment by raffiniert** [comment]
also eagerly awaiting this. Thank you so much for your efforts!

**Comment by hyochan** [comment]
Thank you all for your interest and support. I anticipate that the basic migrations will be completed by mid to late August. As you know, testing in-app purchases has been challenging, with inconsistent results and frequent delays. The testing environment is less than ideal, which has extended the time required. Testing is taking significantly longer than coding.

For this migration, I will only support iOS from StoreKit 2 and Android from BillingClient version 7 onward. Initially, the release will focus on loading products and subscriptions, making purchases, and handling acknowledgments or consumptions. While other functionalities like `getAvailablePurchases` and `getHistories` have also been migrated, they require further testing. I plan to prioritize the main features first and gradually test additional functionalities as time allows. Documentation will be worked on afterward, and I also aim to improve the test codes over time.

Additionally, during the code revision process, I identified several complex or messy areas. This time, I hope to consistently improve and maintain the code without getting tired, working diligently even on weekends. If the in-app functionality is not just a one-time feature but a crucial part of your app that requires continuous stabilization and development, please consider supporting me through [buymeacoffee](https://buymeacoffee.com/hyochan). Your contributions would greatly help maintain this open-source project ☕️

**Reply by farynaio** [reply]
Prioritizing new features instead of preventing regression is bad approach in imo. Our restore purchase feature stopped working after upgrading to `12.16.2`. We had to move to expo new architecture, so we updated react-native-iap to current. At the moment we can't prevent users to use premium features when they canceled subscriptions, as long as they not restart the app.

**Reply by hyochan** [reply]
I understand your point. While we are focused on preventing regression, we are also prioritizing support for the new architecture. Since react-native will eventually be replaced by the new architecture, transitioning to expo-iap will be necessary. Fortunately, as the codebase logic is quite similar, the follow-up process should not be too difficult. Regarding the issue with the restore purchase feature not working after upgrading to `12.16.2`, could you please share any specific details or related issues? This would help us address the problem more effectively. Let us know if you need further assistance!

**Comment by BRC-solutions** [comment]
waiting as well :)

**Comment by n8udd** [comment]
Please don't take this as a "why isn't this ready" kind of comment.

I'm working on my first RN app, and was planning on using Glassfy, but they're shutting down.

I'm now going to use this library, but was wondering whether I'm better off waiting a few weeks untli this is released as an Expo module.

Will it make my life easier in the long term to hold off installing now?

**Comment by fa2id** [comment]
@hyochan Hi, thanks for the library and for supporting the recent upgrades! Since it's almost the end of Aug, could you please let us know when you expect this library to support Expo 51?

We are seeing issues when building the app.

Example:
`* What went wrong:
Could not determine the dependencies of task ':app:compileDebugJavaWithJavac'.
> Could not resolve all dependencies for configuration ':app:debugCompileClasspath'.
   > Could not resolve project :react-native-iap.
     Required by:
         project :app
      > The consumer was configured to find a library for use during compile-time, preferably optimized for Android, as well as attribute 'com.android.build.api.attributes.AgpVersionAttr' with value '8.2.1', attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug', attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm'. However we cannot choose between the following variants of project :react-native-iap:
          - amazonDebugApiElements
          - playDebugApiElements
        All of them match the consumer attributes:
          - Variant 'amazonDebugApiElements' capability MyApp:react-native-iap:unspecified declares a library for use during compile-time, preferably optimized for Android, as well as attribute 'com.android.build.api.attributes.AgpVersionAttr' with value '8.2.1', attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug', attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm':
              - Unmatched attributes:
                  - Provides attribute 'com.android.build.api.attributes.ProductFlavor:store' with value 'amazon' but the consumer didn't ask for it
                  - Provides attribute 'com.android.build.gradle.internal.attributes.VariantAttr' with value 'amazonDebug' but the consumer didn't ask for it
                  - Provides attribute 'store' with value 'amazon' but the consumer didn't ask for it
          - Variant 'playDebugApiElements' capability MyApp:react-native-iap:unspecified declares a library for use during compile-time, preferably optimized for Android, as well as attribute 'com.android.build.api.attributes.AgpVersionAttr' with value '8.2.1', attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug', attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm':
              - Unmatched attributes:
                  - Provides attribute 'com.android.build.api.attributes.ProductFlavor:store' with value 'play' but the consumer didn't ask for it
                  - Provides attribute 'com.android.build.gradle.internal.attributes.VariantAttr' with value 'playDebug' but the consumer didn't ask for it
                  - Provides attribute 'store' with value 'play' but the consumer didn't ask for it
`

**Reply by SterlingVix** [reply]
Facing this exact same issue, which is what brought me here.

I think the move to Expo is a very smart one. Looking forward to a beta branch to test.

**Comment by hyochan** [comment]
### Migration to `expo-iap` and Current Project Status

Hello, `react-native-iap` developers. The migration of `react-native-iap` to `expo-iap` has been delayed due to unexpected workloads, but I have made significant progress during my spare time. While there are still features like hooks that need to be migrated, I have verified that the module works correctly in the projects I have tested.

Please proceed with caution when using this module as it is still under development. I plan to continue working on documentation and completing pending tasks, but updates may not come as quickly as desired. For context, I have not used the IAP package in any app for over six years and have continued maintaining it purely for the sake of keeping the open-source community supported.

Currently, my company is in the process of winding down due to financial difficulties, and I've been experiencing burnout. Through open-sourcing, I've been reflecting on my past while seeking a broader perspective. Although this situation is challenging, it’s not overly dire. My goal was to create a sustainable business through open-source, but I’ve realized how difficult that can be.

The `expo-iap` package is now managed in a separate repository: [https://github.com/hyochan/expo-iap](https://github.com/hyochan/expo-iap). Feel free to test it, and please report any issues in the new repository. The package has also been published separately on npm: [https://www.npmjs.com/package/expo-iap](https://www.npmjs.com/package/expo-iap).

I wanted to share my current status and seek any ideas you might have. Below is a sample code I used while implementing `expo-iap`, which might be helpful when using the module:

```ts
import {
  endConnection,
  finishTransaction,
  getProducts,
  initConnection,
  isProductAndroid,
  isProductIos,
  purchaseErrorListener,
  purchaseUpdatedListener,
  requestPurchase,
} from 'expo-iap';
import type {
  Product,
  ProductPurchase,
  PurchaseError,
} from 'expo-iap/build/ExpoIap.types';
import type {} from 'expo-iap/build/types/ExpoIapAndroid.types';
import {Stack} from 'expo-router';
import {useEffect, useState} from 'react';
import {InteractionManager, View} from 'react-native';
import {t} from '../../../src/STRINGS';
import styled, {css} from '@emotion/native';
import {
  fetchCreatePurchase,
  fetchUserPoints,
} from '../../../src/apis/purchaseQueries';
import {useRecoilValue} from 'recoil';
import {authRecoilState} from '../../../src/recoil/atoms';
import {Button, Icon, Typography, useDooboo} from 'dooboo-ui';
import {showAlert} from '../../../src/utils/alert';

const productSkus = [
  'cpk.points.200',
  'cpk.points.500',
  'cpk.points.1000',
  'cpk.points.5000',
  'cpk.points.10000',
  'cpk.points.30000',
];

const Container = styled.View`
  background-color: ${({theme}) => theme.bg.basic};
  flex: 1;
  align-self: stretch;
`;

const Content = styled.ScrollView`
  padding: 16px;
`;

export default function App() {
  const [isConnected, setIsConnected] = useState(false);
  const [products, setProducts] = useState<Product[]>([]);
  const [userPoints, setUserPoints] = useState(0);
  const {authId} = useRecoilValue(authRecoilState);
  const {theme} = useDooboo();

  useEffect(() => {
    const getUserPoints = async () => {
      const data = await fetchUserPoints(authId!);
      setUserPoints(data || 0);
    };

    authId && getUserPoints();
  }, [authId]);

  useEffect(() => {
    const initIAP = async () => {
      if (await initConnection()) {
        setIsConnected(true);
      }

      const products = await getProducts(productSkus);
      products.sort((a, b) => {
        if (isProductAndroid(a) && isProductAndroid(b)) {
          return (
            parseInt(a?.oneTimePurchaseOfferDetails?.priceAmountMicros || '0') -
            parseInt(b?.oneTimePurchaseOfferDetails?.priceAmountMicros || '0')
          );
        }

        if (isProductIos(a) && isProductIos(b)) {
          return a.price - b.price;
        }

        return 0;
      });
      setProducts(products);
    };

    initIAP();

    return () => {
      endConnection();
    };
  }, []);

  useEffect(() => {
    const purchaseUpdatedSubs = purchaseUpdatedListener(
      (purchase: ProductPurchase) => {
        const ackPurchase = async (purchase: ProductPurchase) => {
          await finishTransaction({
            purchase,
            isConsumable: true,
          });
        };

        InteractionManager.runAfterInteractions(async () => {
          const receipt = purchase && purchase.transactionReceipt;

          if (receipt) {
            const result = await fetchCreatePurchase({
              authId: authId!,
              points: parseInt(purchase?.productId.split('.').pop() || '0'),
              productId: purchase?.productId || '',
              receipt,
            });

            if (result) {
              ackPurchase(purchase);
            }
          }
        });
      },
    );

    const purchaseErrorSubs = purchaseErrorListener((error: PurchaseError) => {
      InteractionManager.runAfterInteractions(() => {
        showAlert(error?.message);
      });
    });

    return () => {
      purchaseUpdatedSubs.remove();
      purchaseErrorSubs.remove();
      endConnection();
    };
  }, [authId]);

  return (
    <Container>
      <Stack.Screen options={{title: t('points.title')}} />
      <View
        style={css`
          align-self: stretch;
          flex-direction: row;
          padding: 8px 16px;
          align-items: center;
          gap: 12px;
          background-color: ${theme.bg.paper};
        `}
      >
        <Typography.Body2
          style={css`
            font-family: Pretendard-Bold;
          `}
        >
          {t('points.myPoints')}
        </Typography.Body2>
        <View
          style={css`
            flex-direction: row;
            align-items: center;
            gap: 4px;
          `}
        >
          <Icon name="Coins" />
          <Typography.Body2
            style={css`
              font-family: Pretendard-Bold;
            `}
          >
            {userPoints}
          </Typography.Body2>
        </View>
      </View>

      <Content>
        {isConnected
          ? products.map((item) => {
              if (isProductAndroid(item)) {
                return (
                  <View
                    key={item.title}
                    style={css`
                      flex: 1;
                      margin-bottom: 8px;
                      flex-direction: row;
                      gap: 12px;
                      justify-content: space-between;
                    `}
                  >
                    <Typography.Body2>
                      <Icon name="Coins" /> {item.title}
                    </Typography.Body2>
                    <Button
                      text={item.oneTimePurchaseOfferDetails?.formattedPrice}
                      onPress={() => {
                        requestPurchase({skus: [item.productId]});
                      }}
                    />
                  </View>
                );
              }

              if (isProductIos(item)) {
                return (
                  <View
                    key={item.id}
                    style={css`
                      flex: 1;
                      margin-bottom: 8px;
                      flex-direction: row;
                      gap: 12px;
                      justify-content: space-between;
                    `}
                  >
                    <Typography.Body2>
                      <Icon name="Coins" /> {item.displayName}
                    </Typography.Body2>
                    <Button
                      text={item.displayPrice}
                      size="small"
                      color="success"
                      styles={{
                        container: css`
                          border-width: 0;
                          width: 88px;
                          padding: 6px 4px;
                        `,
                      }}
                      onPress={() => {
                        requestPurchase({sku: item.id});
                      }}
                    />
                  </View>
                );
              }
            })
          : null}
      </Content>
    </Container>
  );
}
```

Lastly, thank you to [Nami](https://www.namiml.com) who've supported me over the year!

Thank you for your understanding and support.

**Reply by dmathisen** [reply]
Thanks for your amazing contributions. It's much appreciated.
For any new projects, it sounds like the recommendation is to use `expo-iap`, correct?

**Reply by hyochan** [reply]
@dmathisen Yes, that's correct. However, since the progress is slow, I wouldn't recommend using expo-iap just yet. That said, based on my simple tests, it seemed to work well. Moving forward, we need to continue identifying issues and go through the stabilization process.

**Comment by Moi6AP** [comment]
What should I do to run the package with expo go?

**Reply by hyochan** [reply]
You should use [expo dev client](https://docs.expo.dev/versions/latest/sdk/dev-client) because it isn't preinstalled in expo go app.

**Reply by bayramn** [reply]
I just tried it on Expo Managed Flow and it throws error for missing config plugin when trying to build dev-client:
`Failed to resolve plugin for module "react-native-iap" relative to "/Users/User/ExpoApp"`

**Reply by hyochan** [reply]
@bayramn Are you using `react-native-iap` and not `expo-iap`? Also, `react-native-iap` should work though.
Could you kindly tell your progress step by step? I am not sure without further details 🤔

**Reply by Moi6AP** [reply]
Ok, lo probare con expo-dev-client y te dire si me funciona o no, muchas gracias por hacer un paquete para expo 😭

**Reply by mym0404** [reply]
Please write with English in community. Whether using `expo-dev-client` isn't matter but building a standalone build(not expo go) is it. 
Run your build with `expo run:ios` or `expo run:android` and start dev server with `expo start -d` or press `s` in development server menu for switching to development build mode.
Expo go is a development sandbox of Expo and is only able to contain pre-installed native packages.

**Reply by bayramn** [reply]
@hyochan yeah I mixed up both. I removed `expo-iap` and re-installed `react-native-iap`, and package works fine but couldn't make it work after spending all day on it. I looked into other issues and filled the tax and bank details for App Store but still says `Invalid product ID`

        `useEffect(() => {
              const initializeIAP = async () => {
              try {
                await RNIap.initConnection();
                console.log('IAP Initialized');
                // const availableProducts = await RNIap.getSubscriptions({skus: ['safarigpt_weekly_subscription']}); 
                const availableProducts = await RNIap.getProducts({skus: ['weekly_sub']});
                setProducts(availableProducts);
                console.log('Available subscriptions: ', availableSubscriptions);
              } catch (err) {
                console.warn('IAP initialization error: ', err);
              }
            };
            initializeIAP();
            return () => {
              RNIap.endConnection();
            };
        }, []);`

**Comment by hyochan** [comment]
@bayramn 
FYI expo-iap module is tested in https://github.com/crossplatformkorea/CPK/pull/12
Check  `app/(app)/settings/points.tsx`

**Comment by francescoben** [comment]
hi @hyochan , first of all thank you for your excellent work on this library. 👏  I've successfully used it for a mobile application for months. Now I need to rebuild the app and, since I still need to support IAP, I came across this discussion about the future of the project.

I noticed that in September you advised against using it in production environments, and the latest release is from October. What is the current status of the project? Do you plan to work on it in the coming months?

I would need to start using it during Q1 2025, but given its current state, I'm not sure if that's feasible. What would you recommend?

Thanks again!

**Reply by pvinis** [reply]
I'm here as well. I tried expo-iap, and it seems to still have issues with the way it finishes/cancels transactions. I will end up adding revenuecat with their lib for now :(.

**Reply by hyochan** [reply]
Hi @francescoben!

Thanks for the kind words about the library! 👏 Sorry for the late reply—life’s been hectic. `expo-iap` 2.2.0 is out now, tested, and much more stable. I’ll keep working on it in Q1 2025. For your rebuild, I’d recommend using 2.2.0 and testing it out. Should be feasible—let me know if you hit any snags!

Best,
Hyo

**Comment by sumit-1510-sharma** [comment]
Whenever I am trying to make a development build after following the steps in react-native-iap's official documentation like :-

1) install "react-native-iap" with (npm install react-native-iap).
2) add the plugin in app.json or app.config.js.

as I am running `"eas build -p android --profile development"`, it is giving me the following error - 
_"Error: Gradle build failed with unknown error. See logs for the "Run gradlew" phase for more information"_

I tried to use expo-dev-client also and followed the official documentation but I am not able to proceed further this error.

can anyone help me if they figured out something that could be wrong or "react-native-iap" is just not working for now ?, As I've read the whole discussion but this is not mentioned that it doesn't work anymore.

Is expo-iap the way for now ? but it doesn't have any documentation for installation.

**Reply by hyochan** [reply]
Hi! I am back to work.
The documentation is here https://github.com/hyochan/expo-iap/blob/main/iap.md.

**Comment by Carel3DS** [comment]
Hello @hyochan , what's the current status of the migration? Thank you

**Reply by hyochan** [reply]
@Carel3DS  Good question! Please check https://github.com/hyochan/expo-iap/discussions/19#discussioncomment-12492034.

**Comment by Freddy03h** [comment]
Hi,
I understand that Expo is now an official framework, but it's not easy to add Expo modules to an existing bare React Native project (mine started on version 0.13.2). It also doesn't seem like a good idea to include Expo modules just for a single library.

I get that maintaining both the old and new architectures at the same time is complex.
• But why not drop the old architecture in a breaking change?
That would make migration simpler than requiring Expo modules.

I’ve also read about RevenueCat, but migrating to a third-party service and installing an SDK just to access platform APIs doesn’t feel like a real solution either.

Lastly, is this current module compatible with the **Interop Layer** in the new architecture?
At the last App.js Conf, it was mentioned that the Interop Layer is here to stay, so I assume it's still viable to rely on it going forward.

Thank you

**Reply by hyochan** [reply]
Thank you for the very insightful question. I'd like to address the key points you raised.

First, regarding the concern about installing `expo-modules-core` just to use a single module — it's important to clarify what kind of burden this actually entails. In practice, `expo-modules-core` does **not** significantly increase your app size or introduce heavy dependencies. If possible, I'd appreciate it if someone like @evanbacon from the Expo team could confirm this in the thread.

Second, as I understand it, the **Interop Layer** exists to allow legacy modules — which haven't yet adopted the new architecture — to continue working within the TurboModules runtime. It's essentially a bridge designed to buy time for migration. If I'm mistaken, I'd be grateful for clarification. That said, we can’t fully benefit from the new architecture's performance and type safety features through the Interop Layer alone. We still need to write new modules using TurboModules to truly take advantage of the modern system.

For these reasons, I still believe migrating `expo-iap` to the new module system is a strong and forward-looking choice.

**Reply by AdamGerthel** [reply]
> I’ve also read about RevenueCat, but migrating to a third-party service and installing an SDK just to access platform APIs doesn’t feel like a real solution either.

You can't use that library without using RevenueCat's SaaS can you? Which is not free (yes, they have a free tier, but all in all, it's a paid service).

**Reply by radko93** [reply]
I maintain different apps, some with expo modules some without. And while it's not so hard to integrate expo modules, it blocks you from using latest React Native sometimes.

**Reply by hyochan** [reply]
@radko93 I see your point, but I think there might be a bit of confusion here.

You **can** use the latest React Native version with `expo-modules-core` in a `react-native-cli` (bare workflow) project.
`expo-modules-core` by itself does **not** prevent you from upgrading React Native—it’s just a foundational layer that enables Expo-compatible native modules to work outside of the managed workflow.

The limitation you're referring to only applies to fully **managed Expo projects** (using `expo` CLI), which tie to a specific React Native version until the next SDK release.

So to clarify:

* **Bare React Native projects** using `expo-modules-core` (and compatible Expo modules) are free to use the latest React Native versions.
* The version constraint only exists in **managed** projects, not in bare ones using Expo modules.

Kindly correct me if I’m misunderstanding.
@brentvatne @evanbacon

**Reply by radko93** [reply]
@hyochan when you integrate expo modules core you need to alter your native project as well, it might not be compatible between RN versions

**Reply by hyochan** [reply]
@radko93 You raise a valid concern about potential compatibility issues, but I think it's worth noting that while breaking changes *can* happen between RN versions and expo-modules-core, they don't *always* happen. The Expo team generally maintains good compatibility, and when issues do arise, they're typically resolved relatively quickly. That said, you're right that there's always some risk of temporary incompatibility during major RN upgrades.

**Reply by radko93** [reply]
I use expo modules myself so I understand. But right now for example your cannot upgrade to RN 0.80

**Reply by hyochan** [reply]
@radko93 I agree with you these days — you are absolutely right. I actually realized this recently while trying to help out here: https://github.com/hyochan/expo-iap/discussions/150#discussioncomment-14113296. I’m starting to reconsider things from that point 🤔 Thanks!

**Reply by zeeshan-shabbir** [reply]
> I use expo modules myself so I understand. But right now for example your cannot upgrade to RN 0.80

@radko93  you are right i try to install expo modules , using `npx install-expo-modules@latest` which install expo 51 and i have react 0.80+ , then debug whole day to run the build. thanks to your comment we cant use 0.80 unless expo 54 SDK is out.

**Reply by zeeshan-shabbir** [reply]
> @radko93 I agree with you these days — you are absolutely right. I actually realized this recently while trying to help out here: [hyochan/expo-iap#150 (reply in thread)](https://github.com/hyochan/expo-iap/discussions/150#discussioncomment-14113296). I’m starting to reconsider things from that point 🤔 Thanks!

@hyochan , i have project in old Architecture , i  successfully migrated whole project to 0.80, only IAP is pending . should i wait for  expo sdk 54 or find alternate of this  iap. please can you guide me. 
i wasn't not aware that expo wont work with  0.80 i updated whole project to 0.80. thanks for your guidance.

**Reply by hyochan** [reply]
@zeeshan-shabbir I recently ran into this myself and am still figuring it out. I’ll be working on it this month, so stay tuned!

**Reply by hyochan** [reply]
@zeeshan-shabbir I've just made an announcement here #2985! Please follow along!

**Comment by hyochan** [comment]
Documentation is out now for Expo IAP https://expo-iap.hyo.dev
Hope it helps!

**Comment by bjarke-uptime** [comment]
I used to perform receipt validation on iOS by sending the receipt to my backend which then queried Apple's /verifyreceipt endpoint, but 'transactionReceipt' is no longer the same as it was under react-native-iap. Specifically, it is now just a JSON encoded version of the receipt.

Could you make a quick guide on how server side validation is supposed to work with expo-iap?

**Reply by hyochan** [reply]
In StoreKit 2, you’ll need to use `jwsRepresentation` for receipt validation. I’ll be working on this part next month, while focusing on stabilizing our IAP libraries and specs this month. Thanks for your understanding!

**Comment by kushalkulle** [comment]
<img width="904" height="222" alt="Screenshot 2025-08-17 at 12 08 10 PM" src="https://github.com/user-attachments/assets/6a4a2574-3298-4d8a-be04-5c721986bd44" />
Upgraded to the latest version of 2.7.11. And getting this error, need support fixing this.

**Reply by hyochan** [reply]
Related
- https://expo-iap.hyo.dev/docs/guides/faq#what-version-of-xcode-should-i-use-for-ios-development
- https://expo-iap.hyo.dev/docs/guides/faq#build-error-apptransactionid-property-not-found

**Comment by hyochan** [comment]
Maintenance for react-native-iap continues! See #2985 for details
